runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java

```diff
@@ -38,36 +53,117 @@
 public class FlinkExecutableStageFunction<InputT>
     extends RichMapPartitionFunction<WindowedValue<InputT>, RawUnionValue> {
 
+  // Main constructor fields. All must be Serializable because Flink distributes Functions to
+  // task managers via java serialization.
+
   // The executable stage this function will run.
   private final RunnerApi.ExecutableStagePayload stagePayload;
   // Pipeline options. Used for provisioning api.
-  private final Struct pipelineOptions;
+  private final JobInfo jobInfo;
   // Map from PCollection id to the union tag used to represent this PCollection in the output.
   private final Map<String, Integer> outputMap;
+  private final SerializableSupplier<FlinkBundleFactory> bundleFactorySupplier;
+  private final DistributedCachePool.Factory cachePoolFactory;
+  private final FlinkStateRequestHandlerFactory stateHandlerFactory;
+
+  // Worker-local fields. These should only be constructed and consumed on Flink TaskManagers.
+  private transient RuntimeContext runtimeContext;
+  private transient StateRequestHandler stateRequestHandler;
+  private transient StageBundleFactory stageBundleFactory;
+  private transient AutoCloseable distributedCacheCloser;
 
   public FlinkExecutableStageFunction(
       RunnerApi.ExecutableStagePayload stagePayload,
-      Struct pipelineOptions,
-      Map<String, Integer> outputMap) {
+      JobInfo jobInfo,
+      Map<String, Integer> outputMap,
+      SerializableSupplier<FlinkBundleFactory> bundleFactorySupplier,
```

Comment:
Supplier of a factory seems like too many layers of indirection, can the bundle factory itself be serializable?

Reply:
I noticed this comment after leaving another one above (on FlinkBundleFactory).

tl;dr That would be nice in theory, but the FlinkBundleFactory should really only be instantiated on worker nodes and will have machine-local state. I think it would be wrong to make it serializable.

Reply:
Hm I wonder if it makes sense then to encapsulate these three things into a "FlinkExecutableStageContext" and have a ContextFactory, so that at least we avoid the proliferation of factories.